robot/impl/local_robot.go

```diff
@@ -754,7 +756,23 @@ func (r *localRobot) updateWeakDependents(ctx context.Context) {
 	// Track the current value of the resource graph's logical clock. This will
 	// later be used to determine if updateWeakDependents should be called during
 	// getDependencies.
-	r.lastWeakDependentsRound.Store(r.manager.resources.CurrLogicalClockValue())
+	r.lastWeakDependentsMu.Lock()
+	logicalNow := r.manager.resources.CurrLogicalClockValue()
+	if r.lastWeakDependentsRound.Load() < logicalNow {
+		// RSDK-8904: This method can be called concurrently such that resources are `Reconfigure`d
+		// in parallel. This logical clock existed for the purpose of avoiding double-reconfigure
+		// inside `updateWeakDependents` for the same robot `Reconfigure` call.
+		//
+		// Dan: Using the clock in this way will early return the "loser" goroutine. The "loser"
+		// goroutine will proceed despite weak dependents not yet having their `Reconfigure`
+		// called. I expect this is actually safe despite being a surprising lack of guarantee. I'm
+		// concerned about blocking here because I also* expect this logical clock atomic integer
```

Comment:
I think this would be incorrect in the sense that if two components depends on the motion service and are reconfigured at the same time, one might get an updated motion service while the other will early return and thus get a non-updated version.

I don't know if we were excessively concerned about blocking here versus trying to avoid using more locks when introducing the logical clock atomic integer so I would be ok with blocking here (this is basically an implicit dependency that could not be properly expressed). @benjirewis may have more context around the original introduction

Reply:
> one might get an updated motion service while the other will early return and thus get a non-updated version.

I also think this is true unless the first goroutine hols the `lastWeakDependentsMu` until the `Reconfigure` has actually been called on the motion service. But, is that not just an extension of what @dgottlieb is already saying with 

> The "loser" goroutine will proceed despite weak dependents not yet having their `Reconfigure` called. 

I think if we cared about avoiding that situation we would have to block even more and not `r.lastWeakDependsMu.Unlock()` until the end of this function.

> I don't know if we were excessively concerned about blocking here versus trying to avoid using more locks when introducing the logical clock atomic integer so I would be ok with blocking here.

I don't have a ton of context around the original introduction of this field, but I agree with, you, @cheukt. Originally, there was likely very little concurrency leveraging this code, so it was probably assumed that calls to `updateWeakDependents` and `getDependencies` would be sequential. As such, it was probably "good enough" to store a clock value here and assume that the next (sequential) `getDependencies` call would not call `updateWeakDependents` due to the updated `lastWeakDependentsRound` value.

Reply:
fair enough, unless we hold the lock for the entire function and somehow lock for the Reconfiguration of the first component, the second component will never get the "most" up-to-date version of the motion service.

Ok with just leaving the comment there